File: app/MMarket/c/src/data/eval/Result.c
// Copyright 18-Jan-2022 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "data/eval/Result.h"
#include "dmc/DEFS.h"
#include "dmc/js.h"
#include "data/cts.h"

Result *result_new (double assets, double profits, double sales) {
  Result *this = MALLOC(Result);
  this->assets = assets;
  this->profits = profits;
  this->sales = sales;
  return this;
}

double result_eval (Result *this) {
  double max_assets = cts_max_assets_ratio();
  double max_profits = cts_max_profits_avg_ratio();
  double assets = this->assets;
  if (assets > max_assets) assets = max_assets;
  double profits = this->profits;
  if (profits > max_profits) profits = max_profits;

  return
    ( assets * cts_assets_ratio() / max_assets +
      (1 + profits) * cts_profits_avg_ratio() / (1 + max_profits)
    ) / 2
  ;
}

Result *result_sum (Result *r1, Result *r2) {
  return result_new(
    r1->assets + r2->assets,
    r1->profits + r2->profits,
    r1->sales + r2->sales
  );
}

Result *result_div (Result *this, double n) {
  return result_new(
    this->assets / n,
    this->profits / n,
    this->sales /n
  );
}

char *result_to_js (Result *this) {
  return js_wa(achar_new_from(
    js_wd(this->assets),
    js_wd(this->profits),
    js_wd(this->sales),
    NULL
  ));
}

File: app/KtMMarket/go/data/result/result.go
// Copyright 01-Jul-2022 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

// Result data
package result

import (
	"github.com/dedeme/KtMMarket/cts"
	"github.com/dedeme/ktlib/js"
)

// Result data
type T struct {
	// Money
	Assets float64
	// Profits ratio
	Profits float64
	// Number of sales
	Sales float64
}

func New(assets, profits, sales float64) *T {
	return &T{assets, profits, sales}
}

func (r *T) Eval() float64 {
	assets := r.Assets
	if assets > cts.MaxAssets {
		assets = cts.MaxAssets
	}
	profits := r.Profits
	if profits > cts.MaxProfitsAvgRatio {
		profits = cts.MaxProfitsAvgRatio
	}

	return (assets*cts.AssetsRatio/cts.MaxAssets +
		(1+profits)*cts.ProfitsAvgRatio/cts.MaxProfitsAvgRatio) /
		2
}

func (r *T) Sum(r2 *T) *T {
	return New(
		r.Assets+r2.Assets,
		r.Profits+r2.Profits,
		r.Sales+r2.Sales,
	)
}

func (r *T) Div(n int) *T {
	return New(
		r.Assets/float64(n),
		r.Profits/float64(n),
		r.Sales/float64(n),
	)
}

func ToJs(r *T) string {
	return js.Wa([]string{
		js.Wd(r.Assets),
		js.Wd(r.Profits),
		js.Wd(r.Sales),
	})
}
